### ISSUE TITLE
tf.nn.Softmax() crashes if tensor is not two-dimensional 

### ISSUE THREAD

**Issue by vincentleon** [tensorflow/tensorflow#23585]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): ubuntu 16.04 "nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04" docker image and nvidia-docker
- TensorFlow installed from (source or binary): binary (conda install tensorflow-gpu)
- TensorFlow version (use command below): 1.10
- Python version: 3.5
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: cuda 9.2, CuDNN 7.2.1
- GPU model and memory: GeForce GTX 1080 Ti, 11264 MB 

**Describe the current behavior**
I'm working on image segmentation, with 5 classes. After computing logits, I want to compute softmax activations : 
`probabilities = tf.nn.softmax(logits)`
with logits of shape [-1,586,586,5], which shouldn't be a problem since softmax internally reshapes it to a 2D tensor. However, I get the following error.
> Check failed: NDIMS == dims() (2 vs. 4)Asking for tensor of 2 dimensions from a tensor of 4 dimensions

I cannot reproduce the issue outside of docker on ubuntu18.04 and tensorflow 1.11, everything works as expected. This happens at inference time, using a frozen model.

**Describe the expected behavior**
tensor should be reshaped internally and crash should't occur. 

**Quick fix**
` flat_logits = tf.reshape(logits,shape=[-1,n_classes])`
` probabilities = tf.nn.softmax(flat_logits)`
Reshaping the tensor myself prevents the crash from happening.

**Comment by feihugis** [comment]
@vincentleon Could you post an example code for reproducing the problem?

**Comment by harshini-gadige** [comment]
@vincentleon In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here.  If this is no longer an issue, feel free to close.
